Decouple pipeline reductions from final agg reduction

polyfractal opened the pull request:
Historically only two extra activities happened in the final reduction: empty buckets were filled, and pipeline aggs were reduced (since it was the final reduction, this was safe).  Usage of the final reduction is growing however.  Auto-date-histo might need to perform many reductions on final-reduce to merge down buckets, CCS may need to side-step the final reduction if sending to a different cluster, etc

Having pipelines generate their output in the final reduce was convenient, but is becoming increasingly difficult to manage as the rest of the agg framework advances.

This commit decouples pipeline aggs from the final reduction:

1. Introduces a new "top level" reduce, which should be called at the beginning of the reduce cycle (e.g. from the SearchPhaseController)
2. Adds a `materializePipeline()` method to InternalAggs and InternalMultiBucket.  This is essentially the final reduce for pipelines
3. Makes reductions on pipelines a no-op

By separating pipeline reduction into their own set of methods, aggregations are free to use the final reduction for whatever purpose without worrying about generating pipeline results which are non-reducible

Closes https://github.com/elastic/elasticsearch/issues/44914, predecessor PR was https://github.com/elastic/elasticsearch/pull/45359

Commit:
Decouple pipeline reductions from final reduction

Historically only two things happened in the final reduction:
empty buckets were filled, and pipeline aggs were reduced (since it
was the final reduction, this was safe).  Usage of the final reduction
is growing however.  Auto-date-histo might need to perform
many reductions on final-reduce to merge down buckets, CCS
may need to side-step the final reduction if sending to a
different cluster, etc

Having pipelines generate their output in the final reduce was
convenient, but is becoming increasingly difficult to manage
as the rest of the agg framework advances.

This commit decouples pipeline aggs from the final reduction:
1. Introduces a new "top level" reduce, which should be called
at the beginning of the reduce cycle (e.g. from the SearchPhaseController)
2. Adds a `materializePipeline()` method to InternalAggs and
InternalMultiBucket.  This is essentially the final reduce
for pipelines
3. Makes reductions on pipelines a no-op

By separating pipeline reduction into their own set of methods,
aggregations are free to use the final reduction for whatever
purpose without worrying about generating pipeline results
which are non-reducible

Commit:
Remove unnecessary doReduce()

Commit:
Add assertions, javadoc

elasticmachine commented:
Pinging @elastic/es-analytics-geo

polyfractal commented:
Note: the first commit contains the main changes, the second commit (`"Remove unnecessary doReduce()"`) just does a bulk rename since `doReduce()` is no longer needed.  So reviewing the first commit first would be less noisy.

polyfractal commented:
/cc @javanna since you've looked at the final reduce stuff for CCS
/cc @markharwood since I had to make some changes to SigTerms (e.g. the buckets get updated with scores, and I had to adjust some ctors to allow buckets to be recreated with the same score)

Commit:
Merge remote-tracking branch 'origin/master' into materialize_pipeline

not-napoleon commented on server/src/main/java/org/elasticsearch/search/aggregations/InternalAggregations.java:
```diff
@@ -91,10 +92,46 @@ public void writeTo(StreamOutput out) throws IOException {
+        if (context.isFinalReduce()) {
+            List<InternalAggregation> reducedInternalAggs = reduced.getInternalAggregations();
+            reducedInternalAggs = reducedInternalAggs.stream()
+                .map(agg -> agg.materializePipelines(agg, context)).collect(Collectors.toList());
```
nit: I'm not sure if we have a style guide for streams yet, but I'd strongly suggest putting the `.collect()` call on its own line.  Personally, I find streams easier to read if each step gets a line.

not-napoleon commented on server/src/main/java/org/elasticsearch/search/aggregations/InternalMultiBucketAggregation.java:
```diff
@@ -141,6 +142,28 @@ public static int countInnerBucket(Aggregation agg) {
+    }
+
+    private List<B> materializeBuckets(ReduceContext reduceContext) {
+        return getBuckets().stream().map(internalBucket -> {
```
I'm -0 on using streams here.  IMHO, Streams are only good when they serve to enhance readability, and I'm not sure they're doing that here, relative to the equivalent loops.  I'm ok with leaving it this way if you prefer it, but it's not my first choice.  Unless of course you've already looked at the loop based version and decided it's even harder to read, in which case I guess this is the lesser evil.

not-napoleon reviewed (approved):
Someone who knows Significant Terms aggregation better should probably comment on that part, I'm just taking it on faith.  Otherwise this looks good to me.  I left two nits about the use of the Streams API, but they're both a matter of opinion, so don't feel like you have to fix them.

jimczi commented on server/src/main/java/org/elasticsearch/search/aggregations/InternalAggregation.java:
```diff
@@ -126,23 +126,25 @@ public String getName() {
+     * Creates the output from all pipeline aggs that this aggregation is associated with.  Should only
+     * be called after all aggregations have been fully reduced
+     */
+    public InternalAggregation materializePipelines(InternalAggregation reducedAggs, ReduceContext reduceContext) {
```
I wonder if instead of adding `materializePipelines` in the `PipelineAggregator` we could keep `reduce` and just change the name here to be `reducePipeline`. The renaming is not really needed since pipeline and aggregator don't share a base class so there is no confusion ?

jimczi reviewed (commented):
+1 to separate the aggs reduction and the pipeline reduction. However I wonder if all the renamings are needed. Could we just change `InternalAggregations` to separate `reduce` and `reducePipeline` ?

Commit:
Review comments: move back to "reduce" naming

Commit:
Review comments: Stream tweaks

Commit:
More renaming

polyfractal commented:
Review comments addressed.

I renamed the "materialize" stuff back to "reduce", although there's a small hiccup.  Because InternalAggregations's `reduce()` is essentially now `reducePipelines()`, we are left over with the abstract `doReduce()`.

I first renamed that to `reduce()` which makes sense, although it modifies some 70+ files :)

In the second commit I changed it to `doReduce()` which has a much more minimal impact (because all the aggs implemented `doReduce()` abstract method before), although a bit strange conceptually as there isn't a `reduce()` anymore.

I don't have a strong opinion either way, although I'd lean towards renaming to `reduce()` and just accepting that it touches every single aggregation :)

jimczi reviewed (approved):
The change looks good to me.

> I don't have a strong opinion either way, although I'd lean towards renaming to reduce() and just accepting that it touches every single aggregation :)

+1 to move back to `reduce` and no need for another round of review since the change should be straightforward ;).

Commit:
Revert "More renaming"

This reverts commit ba627f61a35089393ebb0b58d7120f1ac333fabb.

polyfractal commented:
@elasticmachine update branch

polyfractal commented:
Oh my goodness, I thought I merged this.  Fixing up conflicts so that it can go in. 

Oops!

Commit:
Merge remote-tracking branch 'origin/master' into materialize_pipeline

Commit:
Merge conflicts

polyfractal commented:
@elasticmachine run elasticsearch-ci/packaging-sample-matrix

Commit:
Merge remote-tracking branch 'origin/master' into materialize_pipeline
